apps/backend/src/app/auth/auth.controller.ts

```diff
@@ -0,0 +1,374 @@
+import {
+  Controller, Post, Body, Logger, HttpCode, HttpStatus, Get, Query, Res,
+  UnauthorizedException
+} from '@nestjs/common';
+import {
+  ApiTags, ApiOkResponse, ApiBadRequestResponse, ApiUnauthorizedResponse, ApiBody, ApiQuery, ApiOperation
+} from '@nestjs/swagger';
+import { Response } from 'express';
+import { OAuth2ClientCredentialsService, ClientCredentialsRequest, ClientCredentialsTokenResponse } from './service/oauth2-client-credentials.service';
+import { KeycloakAuthService, KeycloakUserInfo } from './service/keycloak-auth.service';
+import { AuthService } from './service/auth.service';
+import { CreateUserDto } from '../../../../../api-dto/user/create-user-dto';
+
+@ApiTags('Authentication')
+@Controller('auth')
+export class AuthController {
+  private readonly logger = new Logger(AuthController.name);
+
+  constructor(
+    private readonly oauth2ClientCredentialsService: OAuth2ClientCredentialsService,
+    private readonly keycloakAuthService: KeycloakAuthService,
+    private readonly authService: AuthService
+  ) {}
+
+  /**
+   * Helper method to construct complete callback URI with proper host
+   */
+  private getCallbackUri(): string {
+    const backendUrl = process.env.BACKEND_URL;
+
+    if (backendUrl && backendUrl.startsWith('http')) {
+      return `${backendUrl}api/auth/callback`;
+    }
+    const protocol = process.env.NODE_ENV === 'production' ? 'https' : 'http';
+    const host = process.env.SERVER_NAME || 'localhost';
+    const port = process.env.API_PORT || '3333';
+    const basePath = backendUrl || '';
+    const apiHost = process.env.API_HOST;
+    const isDockerEnv = apiHost && apiHost !== 'localhost';
+
+    if (host === 'localhost' && !isDockerEnv) {
```

Comment:
Dies ist kein PRODUCTION-Code - sollte entfernt werden oder nur in der Dev-Umgebung wirksam werden.
Routing sollte nicht hard-coded in der Anwendung erfolgen, sondern über die 'Config'- bzw. 'Environment'-Variablen geregelt werden.